src/libasr/pass/instantiate_template.cpp

```diff
@@ -401,14 +404,54 @@ class FunctionInstantiator : public ASR::BaseExprStmtDuplicator<FunctionInstanti
         }
         return right;
     }
+
+    /**
+     * @brief Check if the generic function has been instantiated with similar
+     *        arguments. If not, then instantiate a new function.
+     */
+    ASR::symbol_t* get_generic_function(ASR::Function_t *func) {
+        int new_function_num;
+        ASR::symbol_t *t;
+        std::string func_name = func->m_name;
+        if (generic_func_nums.find(func_name) != generic_func_nums.end()) {
+            new_function_num = generic_func_nums[func_name];
+            for (int i=0; i<generic_func_nums[func_name]; i++) {
+                std::string generic_func_name = "__lpython_generic_" + func_name + "_" + std::to_string(i);
+                std::map<std::string, ASR::ttype_t*> subs_check = generic_func_subs[generic_func_name];
+                bool defined = true;
+                for (auto const &subs_check_pair: subs_check) {
+                    if (subs.find(subs_check_pair.first) == subs.end()) {
+                        defined = false;
+                    }
+                    ASR::ttype_t* subs_type = subs[subs_check_pair.first];
+                    ASR::ttype_t* subs_check_type = subs_check_pair.second;
+                    if (!ASRUtils::check_equal_type(subs_type, subs_check_type)) {
+                        defined = false;
+                    }
+                }
+                if (defined) {
+                    t = func_scope->resolve_symbol(generic_func_name);
+                    return t;
+                }
+            }
+        } else {
+            new_function_num = 0;
+        }
+        generic_func_nums[func_name] = new_function_num + 1;
+        std::string new_func_name = "__lpython_generic_" + func_name + "_" + std::to_string(new_function_num);
+        generic_func_subs[new_func_name] = subs;
+        t = ASR::down_cast<ASR::symbol_t>(this->instantiate_Function(func, new_func_name));
+        return t;
+    }
 };
 
+std::map<std::string, int> FunctionInstantiator::generic_func_nums;
+std::map<std::string, std::map<std::string, ASR::ttype_t*>> FunctionInstantiator::generic_func_subs;
```

Comment:
Are these global variables? I think that is asking for trouble regarding parallel compilation and other future improvements. I would not keep any global state.

Reply:
Got it, I was trying to reuse already generated functions. But I think it can easily be done by tracing back to the parent scope.